L74-java-channel-creds.md

```diff
@@ -0,0 +1,411 @@
+nonblocking I/O with a reference-counted, ByteBuffer-backed ByteBuf with an
+event loop. Go and C use a code-based API where credentials implement an
+interface and use a single I/O model. It is not practical to have one
+code-based API to support both OkHttp and Netty, as it would have a
+considerable amount of complexity in the implementations, be hard to optimize,
+and have a wide public API surface. Instead, Java needs a data-based API, where
+configuration values are communicated and each transport can have its own
+implementation. A data-based API can also be used by InProcess and Binder
+transports.
+
+There needs to be a `TlsChannelCredential` which is able to work with both
+OkHttp and Netty transports. This credential needs to live in io.grpc and so
+cannot depend on OkHttp nor Netty. Since it is not feasible for the credential
+to provide the TLS implementation, it will provide configuration. Since new
+configuration will be added over time, there needs to be a way for a transport
+to know that it understands the important parts of the configuration.
+
+To support transport credentials like ALTS and TLS-offloading, there still
+needs to be a way to provide implementations. However, these implementations
+_can_ depend on the specific transport being used. Just the users of those
+implementations should avoid depending _explicitly_ on a particular transport.
+
+The API surface should handle generic and transport-specific credentials in the
+same manner, so the user’s code should not care about the implementation
+details.
+
+It would be possible to use a versioning scheme for `TlsChannelCredentials`.
+Version 1 could be all defaults. Version 2 could have file-based client
+certificates and ca certificates for trust. Version 3 could include ciphers,
+KeyManagers, TrustManagers, and session cache configuration. While versions
+safely allow adding features in the future and is simpler than the Features
+enum approach, it means a consumer that only supports a subset of features at a
+version must check _all_ the features at a particular version to see if they
+are set. Keeping track of which features were available at each version can be
+error-prone. It's also hard to produce a user-meaningful error message, so they
+are aware of which specific configuration is causing the incompatibility.
+
+## Implementation
+
+ejona86 will implement. TLS will initially be bare-bones, with just the
+compatibility checker. Over the long-term TLS will be fleshed out with
+additional configuration. ServerCredentials will be implemented shortly after
+ServerCredentials.
+
+## Open issues (if applicable)
+
+Do we put the constructor in ManagedChannel instead of ManagedChannelBuilder?
+On io.grpc.Grpc? The static methods on ManagedChannelBuilder have been a
+problem, as they are exposed on all subclasses, but don’t operate within the
+scope of the subclass, unless they do.
+
+It would be very convenient to be able to share the TlsChannelCredential API on
```

Comment:
The C++ API has very few of the full range of options defined:
https://grpc.github.io/grpc/cpp/structgrpc__impl_1_1_ssl_credentials_options.html
https://grpc.github.io/grpc/cpp/structgrpc_1_1_ssl_server_credentials_options.html

I think it would be possible to keep our configuration as terse for the most part as well, still using transport-specific APIs for the lots-of-options case. It should also be noticed that C++ allows passing many server certs, such that the only configuration in common between client and server is `pem_root_certs`.

Reply:
Just FWIW, I think in C++ we are trying to add a lot of security-related configurations to tls_credentials_options(https://github.com/grpc/grpc/blob/master/include/grpcpp/security/tls_credentials_options.h), instead of ssl_credentials_options. That's probably the reason why C++ ssl API has very few of options.
Also I think in Go, users can also pass multiple server certs by specifying the `Certificates` field of tls.config(https://golang.org/pkg/crypto/tls/#Config), and use SNI to choose one for the client. I guess it would be great if Java can also support this feature :-)

Reply:
So the new way for users to use TLS is to construct a `TlsCredentialsOptions` and pass it to `TlsCredentials` or `TlsServerCredentials`?

tls.Config in Go supports SNI directly. The Java equivalent would be to let the user pass a SSLContext, but we both know that is a PITA to configure. (It's also not clear to me how we'd consume a raw SSLContext directly in Netty.)

Reply:
Yes, exactly.
+ @jiangtaoli2016 in case he has something to add on.

Reply:
Looking at TlsCredentialsOptions fields:
 * cert_request_type: server-only
 * server_verification_option: server-only
 * key_materials_config: trust and key configuration. Seems like it'd work on both, although it's not clear how multiple keys would be handled on client-side.
 * credentials_reload_config: allows reloading key_materials_config
 * server_authorization_check_config: client-only

So only trust and key configuration is shared today. That's fairly easy to just duplicate and handle twice, as long as we keep the number of representations low (which seems feasible).

Reply:
Yeah. IIRC, for key_material_config, client side is not able to pass in multiple key-identity cert chain pairs; If it does, either the connection would fail, or we pick the first one(I don't remember the details; maybe @jiangtaoli2016 
 or @yihuazhang could add).
From the API's point of view, I think it's probably better to separate at least the  identity key-cert pair part of key_material_config for client and server, to speak the difference out explicitly. For client it would be only one pair; for server it would be multiple pairs, and if multiples are specified, we will use SNI to decide.

Reply:
The details don't really matter, so that's fine.

In short, it seems like "there's very little expected to be shared configuration-wise between client and server." I know there's a laundry list of other TLS features, like which certs and TLS version, but we tend to manage that ourselves and are less likely to need a public, stable API. I am still fine allowing people to configure those things, but they can use the transport-specific APIs.

I'll go ahead and remove the note about sharing configuration. It seems like there's little gain available and probably not worth the confusion cost of mixing both pieces of configuration together.

Reply:
How about creating a limited hierarchy:
```
abstract class TlsCredentials 
class TlsChannelCredentials extends TlsCredentials
class TlsServerCredentials extends TlsCredentials
```
This gives you some sharing.

Reply:
@sanjaypujare, then `TlsChannelCredentials`/`TlsServerCredentials` would not extend `ChannelCredentials`/`ServerCredentials`, as the base credential classes are also abstract classes (not interfaces). We'd need to swap them to interfaces so we could do `class TlsChannelCredentials extends TlsCredentials implements ChannelCredentials`. We could do that, but it encourages more complex hierarchies.